File: domains/query-execution/scheduler/scheduler.ts
#!/usr/bin/env node
/** Declarative query-execution scheduler. */

import { randomUUID } from "node:crypto";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import * as path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

// Paths are computed from the compiled location of this module:
// build/domains/query-execution/scheduler/scheduler.js
//   - repo root  = four levels up from this directory
//   - build root = three levels up (compiled step modules live there)
const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, "..", "..", "..", "..");
const BUILD_ROOT = path.resolve(HERE, "..", "..", "..");
const WORKFLOW = path.join(ROOT, "domains", "query-execution", "scheduler", "workflow.json");
const OUTPUTS = path.join(ROOT, "outputs");

export interface StepOutcomeInit {
  status: string;
  branch?: string;
  context_patch?: Record<string, any>;
  await_step?: string | null;
  payload?: Record<string, any>;
  message?: string;
}

export class StepOutcome {
  status: string;
  branch: string;
  context_patch: Record<string, any>;
  await_step: string | null;
  payload: Record<string, any>;
  message: string;

  constructor(init: StepOutcomeInit) {
    this.status = init.status;
    this.branch = init.branch ?? "next";
    this.context_patch = init.context_patch ?? {};
    this.await_step = init.await_step ?? null;
    this.payload = init.payload ?? {};
    this.message = init.message ?? "";
  }

  static next(patch: Record<string, any> | null = null, branch = "next"): StepOutcome {
    return new StepOutcome({ status: "next", branch, context_patch: patch ?? {} });
  }

  static revise(branch: string, patch: Record<string, any> | null = null, message = ""): StepOutcome {
    return new StepOutcome({ status: "revise", branch, context_patch: patch ?? {}, message });
  }

  static await_input(step_id: string, payload: Record<string, any> | null = null): StepOutcome {
    return new StepOutcome({ status: "await_input", await_step: step_id, payload: payload ?? {} });
  }

  static done(patch: Record<string, any> | null = null): StepOutcome {
    return new StepOutcome({ status: "done", context_patch: patch ?? {} });
  }

  static fail(message: string, patch: Record<string, any> | null = null): StepOutcome {
    return new StepOutcome({ status: "failed", context_patch: patch ?? {}, message });
  }
}

export interface SchedulerStateInit {
  run_id: string;
  current_step: string;
  context?: Record<string, any>;
  status?: string;
  revisions?: Record<string, number>;
  awaiting_step?: string | null;
  await_payload?: Record<string, any>;
  history?: Record<string, any>[];
}

export class SchedulerState {
  run_id: string;
  current_step: string;
  context: Record<string, any>;
  status: string;
  revisions: Record<string, number>;
  awaiting_step: string | null;
  await_payload: Record<string, any>;
  history: Record<string, any>[];

  constructor(init: SchedulerStateInit) {
    this.run_id = init.run_id;
    this.current_step = init.current_step;
    this.context = init.context ?? {};
    this.status = init.status ?? "running";
    this.revisions = init.revisions ?? {};
    this.awaiting_step = init.awaiting_step ?? null;
    this.await_payload = init.await_payload ?? {};
    this.history = init.history ?? [];
  }

  apply(outcome: StepOutcome): void {
    Object.assign(this.context, outcome.context_patch);
    this.history.push({ ...outcome });
    if (outcome.status === "await_input") {
      this.status = "awaiting_input";
      this.awaiting_step = outcome.await_step;
      this.await_payload = outcome.payload;
    } else if (outcome.status === "failed") {
      this.status = "failed";
    } else if (outcome.status === "done") {
      this.status = "completed";
    }
  }
}

export interface WorkflowStep {
  id: string;
  kind?: string;
  run: string;
}

export interface Workflow {
  start: string;
  steps: WorkflowStep[];
  edges?: Record<string, Record<string, string>>;
  backEdges?: Record<string, Record<string, string | number>>;
  [key: string]: any;
}

export type StepRunner = (ctx: Record<string, any>) => StepOutcome;

export class StepScheduler {
  workflow_path: string;
  outputs_dir: string;
  workflow: Workflow;
  steps: Record<string, WorkflowStep>;

  constructor(workflow_path: string = WORKFLOW, outputs_dir: string = OUTPUTS) {
    this.workflow_path = workflow_path;
    this.outputs_dir = outputs_dir;
    this.workflow = JSON.parse(readFileSync(workflow_path, "utf-8")) as Workflow;
    this.steps = {};
    for (const step of this.workflow.steps) {
      this.steps[step.id] = step;
    }
  }

  new_state(context: Record<string, any> | null = null, run_id: string | null = null): SchedulerState {
    return new SchedulerState({
      run_id: run_id ?? randomUUID().replaceAll("-", ""),
      current_step: this.workflow.start,
      context: context ?? {},
    });
  }

  persist(state: SchedulerState): string {
    const run_dir = path.join(this.outputs_dir, state.run_id);
    mkdirSync(run_dir, { recursive: true });
    const file = path.join(run_dir, "state.json");
    writeFileSync(file, JSON.stringify(state, null, 2) + "\n", "utf-8");
    return file;
  }

  resume(run_id: string): SchedulerState {
    const file = path.join(this.outputs_dir, run_id, "state.json");
    const payload = JSON.parse(readFileSync(file, "utf-8")) as SchedulerStateInit;
    return new SchedulerState(payload);
  }

  async run(state: SchedulerState): Promise<SchedulerState> {
    while (state.status === "running") {
      const outcome = await this._run_step(state.current_step, state.context);
      state.apply(outcome);
      if (state.status !== "running") {
        this.persist(state);
        return state;
      }
      state.current_step = this._next_step(state.current_step, outcome, state);
      this.persist(state);
    }
    return state;
  }

  provide_input(state: SchedulerState, payload: Record<string, any>): SchedulerState {
    if (state.status !== "awaiting_input") {
      throw new Error("state is not awaiting input");
    }
    state.context["user_review"] = payload;
    state.status = "running";
    state.awaiting_step = null;
    state.await_payload = {};
    return state;
  }

  private async _run_step(step_id: string, context: Record<string, any>): Promise<StepOutcome> {
    const runner = await this._load_runner(step_id);
    const outcome = runner(context);
    if (!(outcome instanceof StepOutcome)) {
      throw new TypeError(`${step_id}.run(ctx) must return StepOutcome`);
    }
    return outcome;
  }

  private async _load_runner(step_id: string): Promise<StepRunner> {
    const step = this.steps[step_id];
    // workflow.json declares the Python source path of each step; load the
    // compiled .js module from the build tree instead.
    const modulePath = path.join(BUILD_ROOT, step.run.replace(/\.(py|ts)$/, ".js"));
    let module: { run?: StepRunner };
    try {
      module = await import(pathToFileURL(modulePath).href);
    } catch (exc) {
      if ((exc as NodeJS.ErrnoException)?.code === "ERR_MODULE_NOT_FOUND") {
        throw new Error(`cannot load step ${step_id} from ${modulePath}`);
      }
      throw exc;
    }
    if (typeof module.run !== "function") {
      throw new TypeError(`step ${step_id} module at ${modulePath} has no run(ctx) function`);
    }
    return module.run;
  }

  private _next_step(step_id: string, outcome: StepOutcome, state: SchedulerState): string {
    if (this.steps[step_id].kind === "terminal") {
      state.status = "completed";
      return step_id;
    }

    if (outcome.status === "revise") {
      const back = (this.workflow.backEdges ?? {})[step_id] ?? {};
      const target = back[outcome.branch];
      if (target === undefined || target === null) {
        throw new Error(`step ${step_id} cannot revise via branch ${outcome.branch}`);
      }
      const count = (state.revisions[step_id] ?? 0) + 1;
      state.revisions[step_id] = count;
      if (count > Number(back["maxRevisions"] ?? 0)) {
        state.status = "failed";
        return step_id;
      }
      return String(target);
    }

    let branch = outcome.branch;
    if (step_id === "route") {
      branch = "query_path" in state.context ? state.context["query_path"] : branch;
    }
    const target = ((this.workflow.edges ?? {})[step_id] ?? {})[branch];
    if (target === undefined || target === null) {
      throw new Error(`step ${step_id} has no edge for branch ${branch}`);
    }
    return target;
  }
}

async function main(): Promise<number> {
  const scheduler = new StepScheduler();
  const state = scheduler.new_state();
  scheduler.persist(state);
  console.log(`created scheduler state: ${state.run_id}`);
  return 0;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  process.exit(await main());
}

File: domains/query-execution/scheduler/scheduler.py
#!/usr/bin/env python3
"""Declarative query-execution scheduler."""

import importlib.util
import json
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Callable, Optional


ROOT = Path(__file__).resolve().parents[3]
WORKFLOW = Path(__file__).with_name("workflow.json")
OUTPUTS = ROOT / "outputs"


@dataclass
class StepOutcome:
    status: str
    branch: str = "next"
    context_patch: dict[str, Any] = field(default_factory=dict)
    await_step: Optional[str] = None
    payload: dict[str, Any] = field(default_factory=dict)
    message: str = ""

    @classmethod
    def next(cls, patch: Optional[dict[str, Any]] = None, branch: str = "next") -> "StepOutcome":
        return cls(status="next", branch=branch, context_patch=patch or {})

    @classmethod
    def revise(cls, branch: str, patch: Optional[dict[str, Any]] = None, message: str = "") -> "StepOutcome":
        return cls(status="revise", branch=branch, context_patch=patch or {}, message=message)

    @classmethod
    def await_input(cls, step_id: str, payload: Optional[dict[str, Any]] = None) -> "StepOutcome":
        return cls(status="await_input", await_step=step_id, payload=payload or {})

    @classmethod
    def done(cls, patch: Optional[dict[str, Any]] = None) -> "StepOutcome":
        return cls(status="done", context_patch=patch or {})

    @classmethod
    def fail(cls, message: str, patch: Optional[dict[str, Any]] = None) -> "StepOutcome":
        return cls(status="failed", context_patch=patch or {}, message=message)


@dataclass
class SchedulerState:
    run_id: str
    current_step: str
    context: dict[str, Any] = field(default_factory=dict)
    status: str = "running"
    revisions: dict[str, int] = field(default_factory=dict)
    awaiting_step: Optional[str] = None
    await_payload: dict[str, Any] = field(default_factory=dict)
    history: list[dict[str, Any]] = field(default_factory=list)

    def apply(self, outcome: StepOutcome) -> None:
        self.context.update(outcome.context_patch)
        self.history.append(asdict(outcome))
        if outcome.status == "await_input":
            self.status = "awaiting_input"
            self.awaiting_step = outcome.await_step
            self.await_payload = outcome.payload
        elif outcome.status == "failed":
            self.status = "failed"
        elif outcome.status == "done":
            self.status = "completed"


class StepScheduler:
    def __init__(self, workflow_path: Path = WORKFLOW, outputs_dir: Path = OUTPUTS) -> None:
        self.workflow_path = workflow_path
        self.outputs_dir = outputs_dir
        self.workflow = json.loads(workflow_path.read_text(encoding="utf-8"))
        self.steps = {step["id"]: step for step in self.workflow["steps"]}

    def new_state(self, context: Optional[dict[str, Any]] = None, run_id: Optional[str] = None) -> SchedulerState:
        return SchedulerState(
            run_id=run_id or uuid.uuid4().hex,
            current_step=self.workflow["start"],
            context=context or {},
        )

    def persist(self, state: SchedulerState) -> Path:
        run_dir = self.outputs_dir / state.run_id
        run_dir.mkdir(parents=True, exist_ok=True)
        path = run_dir / "state.json"
        path.write_text(json.dumps(asdict(state), indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        return path

    def resume(self, run_id: str) -> SchedulerState:
        path = self.outputs_dir / run_id / "state.json"
        payload = json.loads(path.read_text(encoding="utf-8"))
        return SchedulerState(**payload)

    def run(self, state: SchedulerState) -> SchedulerState:
        while state.status == "running":
            outcome = self._run_step(state.current_step, state.context)
            state.apply(outcome)
            if state.status != "running":
                self.persist(state)
                return state
            state.current_step = self._next_step(state.current_step, outcome, state)
            self.persist(state)
        return state

    def provide_input(self, state: SchedulerState, payload: dict[str, Any]) -> SchedulerState:
        if state.status != "awaiting_input":
            raise ValueError("state is not awaiting input")
        state.context["user_review"] = payload
        state.status = "running"
        state.awaiting_step = None
        state.await_payload = {}
        return state

    def _run_step(self, step_id: str, context: dict[str, Any]) -> StepOutcome:
        runner = self._load_runner(step_id)
        outcome = runner(context)
        if not isinstance(outcome, StepOutcome):
            raise TypeError(f"{step_id}.run(ctx) must return StepOutcome")
        return outcome

    def _load_runner(self, step_id: str) -> Callable[[dict[str, Any]], StepOutcome]:
        step = self.steps[step_id]
        path = ROOT / step["run"]
        spec = importlib.util.spec_from_file_location(step_id.replace(".", "_"), path)
        if spec is None or spec.loader is None:
            raise ImportError(f"cannot load step {step_id} from {path}")
        module = importlib.util.module_from_spec(spec)
        module.StepOutcome = StepOutcome
        spec.loader.exec_module(module)
        return module.run

    def _next_step(self, step_id: str, outcome: StepOutcome, state: SchedulerState) -> str:
        if self.steps[step_id].get("kind") == "terminal":
            state.status = "completed"
            return step_id

        if outcome.status == "revise":
            back = self.workflow.get("backEdges", {}).get(step_id, {})
            target = back.get(outcome.branch)
            if target is None:
                raise ValueError(f"step {step_id} cannot revise via branch {outcome.branch}")
            count = state.revisions.get(step_id, 0) + 1
            state.revisions[step_id] = count
            if count > int(back.get("maxRevisions", 0)):
                state.status = "failed"
                return step_id
            return target

        branch = outcome.branch
        if step_id == "route":
            branch = state.context.get("query_path", branch)
        target = self.workflow.get("edges", {}).get(step_id, {}).get(branch)
        if target is None:
            raise ValueError(f"step {step_id} has no edge for branch {branch}")
        return target


def main() -> int:
    scheduler = StepScheduler()
    state = scheduler.new_state()
    scheduler.persist(state)
    print(f"created scheduler state: {state.run_id}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
